Combine blueprint and commit tx metadata

v0d1ch opened the pull request:
### Why 

We have a user report that the commit auxiliary data hashes are invalid when submitting the commit transaction to blockfrost, responding `ConflictingMetadataHash`.

### What

* Extends the property tests on `commitTx` construction to also check the auxiliary data hash of the resulting blueprint transaction is correct.

* ~~Fixes the auxiliary data hash computation and~~ refactors commit tx construction

* Resolves a warning of missing `txSpendingUTxO` for the `Payment` tx type

* Simplifies definition of `toLedgerTx` and `fromLedgerTx` in `hydra-cardano-api`

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

v0d1ch committed:
Validate blueprint and commit tx metadata

v0d1ch committed:
Add more rich metadata

v0d1ch committed:
Limit the metadata to reasonable sizes

v0d1ch committed:
Add a check for the auxiliary data hashes

v0d1ch committed:
Combine blueprint and commit tx metadata

ch1bo committed:
Implement genMetadata using cardano-ledger functions

The genMatadata' generator seems reasonable for our use case here.

ch1bo committed:
Refactor tests for auxiliary data hash

ch1bo committed:
Refactor how metadata is set on commitTx

ch1bo committed:
Refactor redeemer update in commitTx

This is hopefully a bit clearer to read and maintain going forward.

ch1bo committed:
Resolve compiler warning by defining txSpendinUTxO for Payment

ch1bo committed:
Fix changelog

ch1bo committed:
Improve counter examples in in TxSpec

Reducing noise in the code and providing more counter examples by DRYing
things up a bit.

ch1bo commented:
@v0d1ch Your fix was already correct and I merely refactored and cleaned up the changes. I'd encourage you to double check my refactorings.

@locallycompact @ffakenz You are kindly asked to review the changes as I'm involved now and can't review anymore.

v0d1ch reviewed (commented):
I'd approve but I can't 😀 Probably because I am the author.

ch1bo committed:
Generate blueprints spending from key AND script

This was not covered before and seems like we have failing transactions now.

ffakenz commented on hydra-node/test/Hydra/Chain/Direct/TxSpec.hs:
```diff
@@ -188,68 +208,78 @@ spec =
+                  , conjoin
+                      [ getAuxMetadata blueprintTx `propIsSubmapOf` getAuxMetadata createdTx
+                          & counterexample "Blueprint metadata incomplete"
+                      , propHasValidAuxData blueprintTx
```
`propHasValidAuxData blueprintTx` seems redundant.

my reasoning: 
if getAuxMetadata blueprintTx `propIsSubmapOf` getAuxMetadata createdTx

and

propHasValidAuxData createdTx

does not imply propHasValidAuxData blueprintTx?

ch1bo replied:
No unfortunately not. The validation checks whether the auxiliary data hash and the metadata also line up. The submap check does not involve the auxiliary data.

Checking aux data hash is important as mlabs seems to still have problems with it.

ffakenz commented on hydra-node/test/Hydra/Chain/Direct/TxSpec.hs:
```diff
@@ -329,13 +344,25 @@ genBlueprintTxWithUTxO =
+genMetadata :: Gen TxMetadataInEra
+genMetadata = do
+  genMetadata' @LedgerEra >>= \(ShelleyTxAuxData m) ->
+    pure . TxMetadataInEra . TxMetadata $ fromShelleyMetadata m
```
nice catch: `fromShelleyMetadata` !

ffakenz commented on hydra-node/test/Hydra/Chain/Direct/TxSpec.hs:
```diff
@@ -259,11 +289,10 @@ genBlueprintTxWithUTxO =
  where
   spendingPubKeyOutput (utxo, txbody) = do
-    utxoToSpend <- (genUTxOAdaOnlyOfSize . getPositive) . Positive =<< choose (0, 50)
+    utxoToSpend <- genUTxOAdaOnlyOfSize =<< choose (0, 3)
```
why did we reduce the size?

ch1bo replied:
The code was generating up to 50 inputs, then pruned it down to 1 in `removeRandomInputs`. So, in fact, I increased it from 1 to sometimes `3` ;)

ch1bo committed:
Maybe fix toLedgerTx in presence of aux data changes

We suspect that toLedgerTx is incomplete if the cardano-api Tx was
modified in it's auxiliary data.

v0d1ch committed:
Recalculate aux data hash on conversion and test it

On each conversion we need to recalculate the aux data hash since
cardano-api does not provide a way to set the data hash and on top of
it it seems that it removes it completely.

Add a roundtrip test to assert the metadata is correctly converted.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -142,14 +143,9 @@ instance Arbitrary ClosedThreadOutput where
+-- | The metadata label used for identifying Hydra protocol transactions.
 hydraMetadataLabel :: Word64
-hydraMetadataLabel = 55555
+hydraMetadataLabel = 5
```
Must: I need closure to why this would help the issue we had with auxiliary data hashes?

ch1bo commented on hydra-cardano-api/src/Hydra/Cardano/Api/Tx.hs:
```diff
@@ -234,12 +236,13 @@ toLedgerTx = \case
   -- XXX: The suggested way (by the ledger team) forward is to use lenses to
   -- introspect ledger transactions.
   Ledger.AlonzoTx body wits isValid auxData = ledgerTx
+  body' = body & auxDataHashTxBodyL .~ (hashTxAuxData <$> auxData)
```
This (or above) looks much more likely to be something that helped the root cause..

v0d1ch replied:
we can test if all works without this commit since it didn't fix the original issue according to MLabs.

ch1bo committed:
Define toLedgerTx and fromLedgerTx using ShelleyTx

The cardano-api Tx type is just a wrapper around the ledger and we
should be using that data constructor to do this "conversion".

This also drops a test for "self-healing" auxiliary data hashes, but we
did not require that in the first place: if you modify a transaction
using the ledger-api, make sure to re-compute the auxDataHashTxBodyL.

ch1bo committed:
Update changelog

ch1bo commented:
It turned out that the inconsistency on the auxiliary data hash was on the users side. Nontheless, this PR holds a cleaner implementation of the commit tx construction and some refactorings.
